### PR TITLE
Porting link package to TypeScript.

### LINKED ISSUE #13020: [TS] Rewrite ckeditor5-link into TypeScript
Part of #11704.

### PR DESCRIPTION
**Pull request by niegowski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other (link): Rewrite `ckeditor5-link` to Typescript. Closes #13020.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-image/src/imageupload/imageuploadediting.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ export default class ImageUploadEditing extends Plugin {
 		const conversion = editor.conversion;
 		const fileRepository = editor.plugins.get( FileRepository );
 		const imageUtils = editor.plugins.get( 'ImageUtils' );
-		const imageTypes = createImageTypeRegExp( editor.config.get( 'image.upload.types' )! );
+		const imageTypes = createImageTypeRegExp( editor.config.get( 'image.upload.types' ) as Array<string> ); // TODO temporary
```
Comment (niegowski) [inline packages/ckeditor5-image/src/imageupload/imageuploadediting.ts:78]: This is needed until we fix the image config augmentation.

File: packages/ckeditor5-link/src/linkimageui.ts [hunk 2]
```diff
@@ -108,15 +111,17 @@ export default class LinkImageUI extends Plugin {
...
-	 *
-	 * @private
-	 * @param {module:engine/model/selection~Selection} selection
-	 * @returns {Boolean}
 	 */
-	_isSelectedLinkedImage( selection ) {
+	private _isSelectedLinkedImage( selection: DocumentSelection | Selection ): boolean {
 		const selectedModelElement = selection.getSelectedElement();
-		const imageUtils = this.editor.plugins.get( 'ImageUtils' );
+		const imageUtils = this.editor.plugins.get( 'ImageUtils' ) as ImageUtils;
```
Comment (arkflpc) [inline packages/ckeditor5-link/src/linkimageui.ts:117]:
```suggestion
		const imageUtils: ImageUtils = this.editor.plugins.get( 'ImageUtils' );
```

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 3]
```diff
@@ -239,11 +224,11 @@ export default class LinkActionsView extends View {
 /**
  * Fired when the {@link #editButtonView} is clicked.
  *
- * @event edit
+ * @eventName edit
```
Comment (arkflpc) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:227]: It should be attached to event type.

File: packages/ckeditor5-link/src/ui/linkformview.ts [hunk 4]
```diff
@@ -353,11 +333,11 @@ export default class LinkFormView extends View {
  * Fired when the form view is submitted (when one of the children triggered the submit event),
  * for example with a click on {@link #saveButtonView}.
  *
- * @event submit
+ * @eventName submit
```
Comment (arkflpc) [inline packages/ckeditor5-link/src/ui/linkformview.ts:336]: No event type.

File: packages/ckeditor5-link/src/utils/automaticdecorators.ts [hunk 5]
```diff
@@ -4,45 +4,39 @@
...
 	/**
 	 * Gives information about the number of decorators stored in the {@link module:link/utils~AutomaticDecorators} instance.
 	 *
-	 * @readonly
-	 * @protected
-	 * @type {Number}
+	 * @internal
 	 */
-	get length() {
+	public get length(): number {
```
Comment (arkflpc) [inline packages/ckeditor5-link/src/utils/automaticdecorators.ts:30]: No `_` prefix - is it internal?

**Review by arkflpc — APPROVED** [review]

**Pull request merged**